### PR TITLE
Add basic support for user subuid/subgid management.

### LINKED ISSUE #68199: Add support for subuid and subgid to user module
##### SUMMARY
Subuid's and subgid's are used as part of Linux user namespaces. Heavily relied on by container technologies such as podman/docker.

Adding subuid and subgid support in the user module would make this easier.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
The ansible user module

##### ADDITIONAL INFORMATION
The usermod command supports the following:
```
usermod --add-subuids 200000-201000 --add-subgids 200000-201000 johndoe
```

This results in a corresponding entry in `/etc/subuid` and `/etc/subgid`
```
johndoe:200000:1000
```

It could possibly look, like either this:
```yaml
- name: Add the user 'johndoe' with a specific uid and a primary group of 'admin'
  user:
    name: johndoe
    comment: John Doe
    uid: 1001
    group: admin
    subuids: 200000-201000
    subgids: 200000-201000
```
or this:
```yaml
- name: Add the user 'johndoe' with a specific uid and a primary group of 'admin'
  user:
    name: johndoe
    comment: John Doe
    uid: 1001
    group: admin
    subuid: 200000:1000
    subgid: 200000:1000
```
Depending on why you're using it, there may need to be a handler run to execute some commands, similar to the systemd daemon needing a reload after modifying a service file.  I'm just not 100% sure on what all the possible combinations of a user task being "changed" are that would require this handler to be triggered.  IE, updating a comment or group would not.

### PR DESCRIPTION
**Pull request by gordonmessmer**
##### SUMMARY

Fixes #68199 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
user

##### ADDITIONAL INFORMATION

This addition does not support arbitrary static assignments of subuid or subgid, but does support ensuring that a range is present for a user.

For example:

```
- name: set subid for gordon
  user: name=gordon subuid=present subgid=present
```

### CODE REVIEW

**Comment by PayBas** [comment]
> This addition does not support arbitrary static assignments of subuid or subgid, but does support ensuring that a range is present for a user.

I think this is by far the most common (and useful) use-case. I don't think many people will care that the exact ranges aren't deterministic.

**Comment by gordonmessmer** [comment]
65536 is used as a default value in case login.defs does not specify a count.  The block immediately following that value will load the values for min, max, and count sub-ids from login.defs.

**Comment by samdoran** [comment]
The min, max, and count values should be parameters as well. This also needs integration tests and a changelog.

**Comment by gordonmessmer** [comment]
The integration test will need a little more work.  For systems that support subuid and subgid, ranges are assigned automatically when users are added.  Currently only "subuid: present" is supported, and I expect that the module will have to add support for at least "subuid: absent" in order to get meaningful test results.

**Comment by gordonmessmer** [comment]
@samdoran Integration tests and changelog have been added.

I'm not sure it makes sense to add min/max/count as user-specified values, since those are system-dependent values loaded from /etc/login.defs.

As for the rationale behind this implementation:  Systems that support subuid and subgid will assign such a range when new local users are created.  However, systems that use non-local users (such as FreeIPA) do not currently have a centralized means of distributing or assigning such ranges.  Those features are being discussed upstream, but there is no ETA for an implementation.  In the mean time, indefinitely, organizations with central user management which want to use rootless containers must assign subuid/subgid ranges on each system individually.

This implementation follows the default behavior of useradd, assigning a subuid range for users based on the values specified in login.defs.

The underlying system supports storing multiple ranges in the subuid/subgid file, so in order to properly support the specification of arbitrary static ranges in an Ansible playbook, the user module would need to be able to detect overlap and intersection with the existing specification which might provide multiple ranges.  I believe this would involve rather a lot of code in addition to the patch that I've proposed in order to correctly provide an idempotent module.  I don't think the value of that addition justifies the effort that would be required.  For my own purposes, I only need to ensure that non-local users have a range that's consistent with the range that would be added if they had been created as a local user.

File: lib/ansible/modules/user.py [hunk 1]
```diff
@@ -250,6 +250,18 @@
             - Supported on Linux only.
         type: int
         version_added: "2.11"
+    subuid:
+        description:
+            - UID range allocated for use by newuidmap.
+            - Supported on Linux only.
+            - Currently only C(subuid=present) and C(subuid=absent) are supported.
```
Comment (felixfontein) [inline lib/ansible/modules/user.py:257]:
```suggestion
            - Currently only I(subuid=present) and I(subuid=absent) are supported.
```
See https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#linking-and-other-format-macros-within-module-documentation

(Same below.)

File: lib/ansible/modules/user.py [hunk 2]
```diff
@@ -250,6 +250,18 @@
...
+        type: str
```
Comment (felixfontein) [inline lib/ansible/modules/user.py:258]:
```suggestion
        type: str
        version_added: 2.12
```
(Same below.)

File: lib/ansible/modules/user.py [hunk 3]
```diff
@@ -433,6 +445,16 @@
   returned: When user exists
   type: int
   sample: 20
+subuid:
+  description: UID range allocated for use by newuidmap
+  returned: When subuid range has been allocated
+  type: str
+  sample: 100000-165535
```
Comment (felixfontein) [inline lib/ansible/modules/user.py:452]:
```suggestion
  sample: 100000-165535
  version_added: 2.12
```
(Same below.)

Also, why not return this in a way that can be processed without parsing it first? I.e. provide start and end (or start, end and count) as separate values, or a dictionary with fitting keys?

File: lib/ansible/modules/user.py [hunk 4]
```diff
@@ -3018,6 +3122,8 @@ def main():
             login_class=dict(type='str'),
             password_expire_max=dict(type='int', no_log=False),
             password_expire_min=dict(type='int', no_log=False),
+            subuid=dict(type='str'),
```
Comment (felixfontein) [inline lib/ansible/modules/user.py:3125]: I would add:
```suggestion
            # following options are specific to Linux
            subuid=dict(type='str'),
```

**Review by felixfontein — COMMENTED** [review]
How about making the new options dictionaries with a suboption `state`, so that something like https://github.com/ansible/ansible/issues/68199#issuecomment-873434876 could be added later? Otherwise we'd have to invent syntaxes for providing a range (that would be easy), only a count, or start + count.

**Comment by secustor** [comment]
We are facing this issue too. Is there anything which we can do to move this along?

**Pull request closed without merging**